synapse/storage/state.py

```diff
@@ -185,7 +185,19 @@ def get_state_groups(self, room_id, event_ids):
 
     @defer.inlineCallbacks
     def _get_state_groups_from_groups(self, groups, types):
-        """Returns dictionary state_group -> (dict of (type, state_key) -> event id)
+        """Returns the state groups for a given set of groups, filtering on
+        types of state events.
+
+        Args:
+            groups(list[int]): list of state group IDs to query
+            types(list[str|None, str|None])|None: List of 2-tuples of the form
+                (`type`, `state_key`), where a `state_key` of `None` matches all
+                state_keys for the `type`. Presence of type of `None` indicates
```

Comment:
> Presence of type of `None` indicates that types not in the list should not be filtered out

I'm afraid I totally failed to grok this (which you have dutifully copied throughout the PR, thank you) without reading the code. I think it needs more explanation: "An entry of `(None, None)` is treated as a special case which means that any state events whose `type` does not appear anywhere else in the list will be included in the returned value", or words to that effect.

I'm also led to wonder if we shouldn't just make `include_other_types` another separate argument rather than this rather obscure special case.

Finally, I'm concerned that this API has rather surprising behaviour. For instance, suppose I wanted to get the list of all state events, except `m.room.member` which I want to filter by a list of users:

```
def get_limited_state(groups, users):
    _get_state_groups_from_groups(
        groups,
        [('m.room.member', u) for u in users],
        include_other_types=True,
   )
```

Now suppose I call that with an empty set of `users`. The expected behaviour is to get *no* `m.room.member` events; the *actual* behaviour is that I get *all* of them. Synapse has enough footguns without adding more.

An ideal solution might be for `types` to instead be a `dict[str, iterable[str]|None]` mapping from `type` to a set of state_keys, but that might require quite a lot of refactoring throughout the codebase. As a compromise: instead of the `(None, None)` special case (or a separate `include_other_types` param) add a `filtered_types` param (which would be set to `['m.room.member']` in the above usecase)?

Reply:
have adopted `filtered_types` as suggested; it definitely feels better - thanks.